File: capi/capi.c
#include "capi.h"
#include <stdio.h>
#include <stdlib.h>
int pass_by_val(int one, double two)
{
  printf("%d,%f\n",one,two);
  return 0;
}
int pass_by_ptr(int * one, double * two)
{
  *one += 1;
  *two += 1;
  printf("%d,%f\n",*one,*two);
  return 0;
}
struct Node
{
  Node * l;
  Node * r;
  int val;
};
Node * create_node(double v, Node * l, Node * r)
{
  Node * rslt = malloc(sizeof(*rslt));
  rslt->l = l;
  rslt->r = r;
  rslt->val = v;
  printf("N @ %d: value %f, left %p, right %p, created\n",rslt,v,l,r);
  return rslt;
}
void destroy_node(Node * n)
{
  free(n);
}
void add_left(Node * n, Node * l)
{
  n->l = l;
  printf("N @ %d: add left @ %p\n",n,l);
}
void add_right(Node * n, Node * r)
{
  n->r = r;
  printf("N @ %d: add right @ %p\n",n,r);
}
Node * get_left(Node * n)
{
  printf("N @ %d: get left: %p\n",n,n->l);
  return n->l;
}
Node * get_right(Node * n)
{
  printf("N @ %d: get right: %p\n",n,n->r);
  return n->r;
}
int get_node_val(Node * n)
{
  printf("N @ %d: get val: %f\n",n,n->val);
  return n->val;
}

File: fsrc/capi.f90
module capi_mod
  interface
     integer(c_int) function pass_by_val(one,two) bind(c)
       use iso_c_binding
       implicit none
       integer(c_int), value :: one
       real(c_double), value :: two
     end function pass_by_val
     integer(c_int) function pass_by_ptr(one,two) bind(c)
       use iso_c_binding
       implicit none
       integer(c_int) :: one
       real(c_double) :: two
     end function pass_by_ptr
     type(c_ptr) function create_node(v,l,r) bind(c)
       use iso_c_binding
       implicit none
       real(c_double), value :: v
       type(c_ptr), value    :: l
       type(c_ptr), value    :: r
     end function create_node
     subroutine destroy_node(n) bind(c)
       use iso_c_binding
       implicit none
       type(c_ptr), value :: n
     end subroutine destroy_node
     subroutine add_left(n,l) bind(c)
       use iso_c_binding
       implicit none
       type(c_ptr), value :: n
       type(c_ptr), value :: l
     end subroutine add_left
     subroutine add_right(n,r) bind(c)
       use iso_c_binding
       implicit none
       type(c_ptr), value :: n
       type(c_ptr), value :: r
     end subroutine add_right
     type(c_ptr) function get_left(n) bind(c)
       use iso_c_binding
       implicit none
       type(c_ptr), value :: n
     end function get_left
     type(c_ptr) function get_right(n) bind(c)
       use iso_c_binding
       implicit none
       type(c_ptr), value :: n
     end function get_right
     integer(c_int) function get_node_val(n) bind(c)
       use iso_c_binding
       implicit none
       type(c_ptr), value :: n
     end function get_node_val
  end interface
contains
end module capi_mod
